### PR TITLE
Add Swift tests for Messaging

### PR DESCRIPTION
**Pull request by chliangGoogle**
Adding Swift API and Integration tests from Messaging

#no-changelog

### CODE REVIEW

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        return
+      }
+      messaging.unsubscribe(fromTopic: "cat_video") { error in
+        XCTAssertNil(error)
+        expectation.fulfill()
+      }
+      wait(for: [expectation], timeout: 5)
+    }
+
+    func assertDefaultToken() {
```
Comment (paulb777) [inline FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:70]: Please add the async variation of the test like `testScalar` and `testScalarAsync` at https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseFunctions/Tests/SwiftIntegration/IntegrationTests.swift#L97

File: FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  if let _ = messaging.apnsToken {}
+
+  let apnsToken = Data()
+  messaging.setAPNSToken(apnsToken, type: .prod)
+
+  let topic = "cat_video"
+  messaging.subscribe(toTopic: topic)
+  messaging.unsubscribe(fromTopic: topic)
+  messaging.appDidReceiveMessage([:])
+}
```
Comment (paulb777) [inline FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift:36]: I don't see the public APIs from FIRMessagingExtensionHelper.h covered.

File: FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import FirebaseCore
+import FirebaseMessaging
+
+// This file is a build-only test for the public Messaging Swift APIs not
+// exercised in the integration tests.
+func apis() {
+  let messaging = Messaging.messaging()
+
+  Messaging.serviceExtension()
```
Comment (paulb777) [inline FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift:25]: Based on test build failure, it looks like an `available` check is needed.

**Review by paulb777 — COMMENTED** [review]
Good progress! 

Here's an example of setting up a GoogleService-Info.plist for integration tests - https://github.com/firebase/firebase-ios-sdk/blob/master/scripts/build.sh#L528

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift [hunk 4]
```diff
@@ -97,6 +96,30 @@
...
+        object: nil,
+        handler: nil)
+
+      let testDelegate = fakeAppDelegate()
+      messaging?.delegate = testDelegate
+      testDelegate.delegateIsCalled = false
+
+      guard let messaging = self.messaging else {
+        return
+      }
```
Comment (ncooke3) [inline FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift:114]: I think it'd be helpful to ensure that the `messaging` property is non-nil by the time each test case runs. This prevents need for code in each test case that works around the fact that `messaging` is an optional property.

- Declare the `messaging` as explicitly unwrapped
- Use [setUpWithError](https://developer.apple.com/documentation/xctest/xctest/3521150-setupwitherror?language=objc) API to assign it

Here's what the refactor would look like:
```diff
 class FIRMessagingTokenRefreshTests: XCTestCase {
    var app: FirebaseApp!
-    var messaging: Messaging?
+    var messaging: Messaging!
    override class func setUp() {
      FirebaseApp.configure()
    }
-    override func setUp() {
+   override func setUpWithError() throws {
      messaging = Messaging.messaging()
+    messaging = try XCTUnwrap(Messaging.messaging())
    }
```

With this refactor, the optional related code in the tests can be removed. For this case:

```suggestion
      let testDelegate = fakeAppDelegate()
      messaging.delegate = testDelegate
      testDelegate.delegateIsCalled = false
```

The same refactor can be applied to FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift to simplify things as well.

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift (hunk 1 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:70]: > 



> Good progress!
> 
> Here's an example of setting up a GoogleService-Info.plist for integration tests - https://github.com/firebase/firebase-ios-sdk/blob/master/scripts/build.sh#L528

@paulb777 Is the build script for presubmit tho? The plist file is not associated with test bundle locally when running  pod gen FirebaseMessaging.podspec

Comment (paulb777) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:70]: @chliangGoogle For running locally, you either need to run the ruby command or manually add `GoogleService-Info.plist` to the  AppHost-FirebaseStorageMessaging... target in the Xcode UI

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift (hunk 4 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift:114]: Done. Thanks!

File: FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift (hunk 3 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift:25]: Done.

File: FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift (hunk 2 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift:36]: Good catch. Done.

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift (hunk 1 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:66]: @paulb777 Can you give more context on the async vatiation test? I do have async tests in integration. What are the differences here?

Comment (paulb777) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:66]: There should be tests to verify the auto-generated `async` versions of the APIs that have a completion handler like the difference between the `testScalar` and `testScalarAsync` linked above.

Comment (chliangGoogle) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:66]: Oh this is for the async/await! Gotcha!

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// macOS requests a user password when accessing the Keychain for the first time,
+// so the tests may fail. Disable integration tests on macOS so far.
+// TODO: Configure the tests to run on macOS without requesting the keychain password.
```
Comment (paulb777) [inline FirebaseMessaging/Tests/IntegrationTests/FIRMessagingPubSubTest.swift:19]: Might be solvable with a similar approach as #9102 for FIS.

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift [hunk 6]
```diff
@@ -35,14 +34,16 @@
 
   class FIRMessagingTokenRefreshTests: XCTestCase {
     var app: FirebaseApp!
-    var messaging: Messaging?
+    var messaging: Messaging!
```
Comment (paulb777) [inline FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift:37]: Is the force unwrap removable by changing the assignment below to a `guard let`?

File: FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    let serviceExtension = Messaging.serviceExtension()
+    let content = UNMutableNotificationContent()
+    serviceExtension.populateNotificationContent(content) { content in
+    }
+    serviceExtension.exportDeliveryMetricsToBigQuery(withMessageInfo: [:])
+  }
+}
+
+@available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+func apiAsync() async {
```
Comment (paulb777) [inline FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift:47]: Make this `async throws` and eliminate the `do` and `catch`es below.

**Review by paulb777 — APPROVED** [review]

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift (hunk 6 above)
Comment (ncooke3) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift:37]: No, you can only remove it if you either assign a value when it's declared or add a designated initializer to initialize it.

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift [hunk 8]
```diff
@@ -97,6 +95,30 @@
...
+        object: nil,
+        handler: nil)
+
+      let testDelegate = fakeAppDelegate()
+      messaging?.delegate = testDelegate
+      testDelegate.delegateIsCalled = false
+
+      guard let messaging = self.messaging else {
+        return
+      }
```
Comment (ncooke3) [inline FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift:113]: Based on previous refactor you did, can this now be?:
```suggestion
      messaging.delegate = testDelegate
      testDelegate.delegateIsCalled = false
```

**Review by ncooke3 — APPROVED** [review]

File: FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift (hunk 7 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/UnitTestsSwift/FIRMessagingAPITest.swift]: Nice. much cleaner.

File: FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift (hunk 8 above)
Comment (chliangGoogle) [reply FirebaseMessaging/Tests/IntegrationTests/FIRMessagingTokenRefreshTests.swift]: Done!

**Review by paulb777 — APPROVED** [review]
Thanks!

**Pull request merged**